ReactiveCocoa/Swift/Lifetime.swift

```diff
@@ -0,0 +1,36 @@
+import enum Result.NoError
+
+/// Provides a signal that completes when the lifetime object is deinitialized.
+///
+/// When assigned to a property of another object, provides a hook to
+/// observe when that object goes out of scope.
+public final class Lifetime {
+	/// A signal that sends a Completed event when the lifetime ends.
+	public let ended: Signal<(), NoError>
+
+	private let endedObserver: Observer<(), NoError>
+
+	public init() {
+		(ended, endedObserver) = Signal.pipe()
+	}
+
+	deinit {
+		endedObserver.sendCompleted()
+	}
+}
+
+private var lifetimeKey: UInt8 = 0
+
+extension NSObject {
+	/// Returns a lifetime that ends when the receiver is deallocated.
+	@nonobjc public var lifetime: Lifetime {
+		if let lifetime = objc_getAssociatedObject(self, &lifetimeKey) as! Lifetime? {
```

Comment:
Apparently we could have a race condition here. This would need to be guarded by the `objc_sync_*` pair or some kind of locks.